datafusion/core/src/execution/memory_manager/mod.rs

```diff
@@ -97,339 +89,154 @@ impl MemoryManagerConfig {
     /// Create new memory manager based on the configuration
-    #[allow(clippy::mutex_atomic)]
     pub fn new(config: MemoryManagerConfig) -> Arc<Self> {
-        let pool_size = config.pool_size();
-
         match config {
             MemoryManagerConfig::Existing(manager) => manager,
-            MemoryManagerConfig::New { .. } => {
+            MemoryManagerConfig::New {
+                max_memory,
+                memory_fraction,
+            } => {
+                let pool_size = (max_memory as f64 * memory_fraction) as usize;
                 debug!(
                     "Creating memory manager with initial size {}",
                     human_readable_size(pool_size)
                 );
 
                 Arc::new(Self {
-                    requesters: Arc::new(Mutex::new(HashSet::new())),
-                    pool_size,
-                    requesters_total: Arc::new(Mutex::new(0)),
-                    trackers_total: AtomicUsize::new(0),
-                    cv: Condvar::new(),
+                    state: Arc::new(MemoryManagerState {
+                        pool_size,
+                        used: AtomicUsize::new(0),
+                    }),
                 })
             }
         }
     }
 
-    fn get_tracker_total(&self) -> usize {
-        self.trackers_total.load(Ordering::SeqCst)
-    }
-
-    pub(crate) fn grow_tracker_usage(&self, delta: usize) {
-        self.trackers_total.fetch_add(delta, Ordering::SeqCst);
-    }
-
-    pub(crate) fn shrink_tracker_usage(&self, delta: usize) {
-        let update =
-            self.trackers_total
-                .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |x| {
-                    if x >= delta {
-                        Some(x - delta)
-                    } else {
-                        None
-                    }
-                });
-        update.unwrap_or_else(|_| {
-            panic!(
-                "Tracker total memory shrink by {} underflow, current value is ",
-                delta
-            )
-        });
+    /// Returns the maximum pool size
+    ///
+    /// Note: this can be less than the amount allocated as a result of [`MemoryManager::allocate`]
+    pub fn pool_size(&self) -> usize {
+        self.state.pool_size
     }
 
-    /// Return the total memory usage for all requesters
-    pub fn get_requester_total(&self) -> usize {
-        *self.requesters_total.lock()
+    /// Returns the number of allocated bytes
+    ///
+    /// Note: this can exceed the pool size as a result of [`MemoryManager::allocate`]
+    pub fn allocated(&self) -> usize {
+        self.state.used.load(Ordering::Relaxed)
     }
 
-    /// Register a new memory requester
-    pub(crate) fn register_requester(&self, requester_id: &MemoryConsumerId) {
-        self.requesters.lock().insert(requester_id.clone());
+    /// Returns a new empty allocation identified by `name`
+    pub fn new_tracked_allocation(&self, name: String) -> TrackedAllocation {
```

Comment:
I think it is worth leaving some sort of `requester_id` as part of this API so that it is possible to extend in the future so that the first spilling operator doesn't end up with all the memory, for example.

Reply:
I think I would rather punt on this until we have such a system in place, I am skeptical that such a system would make use of IDs a not just an Arc